content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,260 @@
+[limitations](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos#reasons-for-changing-the-formula-of-memoryhigh-calculation-in-alpha-v127) 
+around the formula for calculating memory.high were identified. These limitations are 
+addressed in Kubernetes v1.27.
+
+## Background
+
+Kubernetes allows you to optionally specify how much of each resources a container needs
+in the Pod specification. The most common resources to specify are CPU and Memory. 
+
+For example, a Pod spec with container resources could look like as follows:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+  - name: nginx
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "64Mi"
+        cpu: "500m"
+```
+
+* `spec.containers[].resources.requests`
+
+  When you specify the resource request for containers in a Pod, the [kube-scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/#kube-scheduler)
+  uses this information to decide which node to place the Pod on. Each node has a 
+  maximum capacity for each of the resource types: the amount of CPU and memory it 
+  can provide for Pods. The scheduler ensures that, for each resource type, the sum
+  of the resource requests of the scheduled containers is less than the capacity of
+  the node.
+
+* `spec.containers[].resources.limits`
+
+  When you specify the resource limit for containers in a Pod, the kubelet enforces 
+  those limits so that the running containers are not allowed to use more of those 
+  resources than the limits you set.
+
+When a node with enough capacity to serve the requested resources has been identified, 
+the kubelet starts the containers. When the kubelet starts a container as a part of a Pod, 
+kubelet passes the container's requests and limits for CPU and memory to the container 
+runtime. On Linux, the container runtime typically configures kernel cgroups as follows:
+
+* CPU limit
+
+  CPU is considered a compressible resource. Linux kernel periodically checks to see if
+  the CPU limit for a cgroup is exceeded. If your app starts hitting your CPU limits, 
+  Kubernetes starts throttling your container, giving your app potentially worse 
+  performance. However, it won’t be terminated. That is what "compressible" means.
+  
+* CPU request
+
+  The CPU request typically defines a weighting. If several different containers (cgroups)
+  want to run on a contended system, workloads with larger CPU requests are allocated more 
+  CPU time than workloads with small requests.
+  
+* Memory limit
+  
+  The memory limit specifies the memory limit, beyond which if the container tries to allocate 
+  more memory, Linux kernel will terminate it with an OOM (Out of Memory) kill.
+
+  In cgroup v1, `memory.limit_in_bytes` interface is used to set the memory usage limit.
+  However, unlike CPU memory can't be compressed as cgroup v1 doesn't provide any interface
```

Comment:
I would rephrase this a bit -- memory compression and memory throttling are different topics, so I would suggest to avoid mixing them. I think the point here we want to make is that an app in cgroupv1 using memory would be fine... and then as soon as hitting the limit get killed right away. There is no way to provide an early warning or attempt to reclaim memory usage before it would be too late and the containerd would be OOMd.